torchvision/transforms/v2/functional/_utils.py

```diff
@@ -1,8 +1,167 @@
+import functools
+import inspect
+import warnings
 from typing import Any
 
 import torch
+from torchvision import datapoints
 from torchvision.datapoints._datapoint import Datapoint
 
 
 def is_simple_tensor(inpt: Any) -> bool:
     return isinstance(inpt, torch.Tensor) and not isinstance(inpt, Datapoint)
+
+
+_KERNEL_REGISTRY = {}
+
+
+def _kernel_wrapper_internal(dispatcher, kernel):
+    dispatcher_params = list(inspect.signature(dispatcher).parameters)[1:]
+    kernel_params = list(inspect.signature(kernel).parameters)[1:]
+
+    needs_args_kwargs_handling = kernel_params != dispatcher_params
+
+    kernel_params = set(kernel_params)
+    explicit_metadata = {
+        input_type: available_metadata & kernel_params
+        for input_type, available_metadata in [(datapoints.BoundingBoxes, {"format", "canvas_size"})]
+    }
+
+    @functools.wraps(kernel)
+    def wrapper(inpt, *args, **kwargs):
+        input_type = type(inpt)
+
+        if needs_args_kwargs_handling:
+            # Convert args to kwargs to simplify further processing
+            kwargs.update(dict(zip(dispatcher_params, args)))
+            args = ()
+
+            # drop parameters that are not relevant for the kernel, but have a default value
+            # in the dispatcher
+            for kwarg in kwargs.keys() - kernel_params:
+                del kwargs[kwarg]
+
+            # add parameters that are passed implicitly to the dispatcher as metadata,
+            # but have to be explicit for the kernel
+            for kwarg in explicit_metadata.get(input_type, set()):
+                kwargs[kwarg] = getattr(inpt, kwarg)
```

Comment:
The more logic/magic there is here, the more I'm tempted to just write manual kernels. All of this happens at runtime right? For every single kernel call, on every input? I'd be more comfortable looking at perf benchmarks before we get this in

Reply:
We decided offline, that we can always go for this later. I'll go with the manual wrappers for now.